src/blackbox.py

```diff
@@ -0,0 +1,270 @@
+#!/usr/bin/env python3
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Workload manager for Blackbox Exporter."""
+
+import logging
+import re
+import time
+import urllib.error
+import urllib.parse
+import urllib.request
+from typing import Optional
+
+import yaml
+from ops.framework import Object
+from ops.pebble import (  # type: ignore
+    ChangeError,
+    Layer,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class WorkloadManagerError(Exception):
+    """Base class for exceptions raised by WorkloadManager."""
+
+
+class ConfigUpdateFailure(WorkloadManagerError):
+    """Custom exception for failed config updates."""
+
+
+class ContainerNotReady(WorkloadManagerError):
+    """Raised when an operation is run that presumes the container being ready.."""
+
+
+class WorkloadManager(Object):
+    """Workload manager for blackbox exporter."""
+
+    _layer_name = _service_name = "blackbox-exporter"
+    _exe_name = "blackbox_exporter"
+
+    def __init__(
+        self,
+        charm,
+        *,
+        container_name: str,
+        port: int,
+        web_external_url: str,
+        config_path: str,
+    ):
+        # Must inherit from ops 'Object' to be able to register events.
+        super().__init__(charm, f"{self.__class__.__name__}-{container_name}")
+
+        self._unit = charm.unit
+
+        self._service_name = self._container_name = container_name
+        self._container = charm.unit.get_container(container_name)
+
+        self.api = BlackboxExporterApi(endpoint_url=charm._external_url)
```

Comment:
Do we need an HTTP client for BBE?

Reply:
**Need** is arguable, but I think the abstraction is useful to have; it's handy to reload the configuration, and could be useful for future features as well.